web/viewer.js: Persist the state of sidebar

ankitaggarwal011 opened the pull request:
Persist the state of content sidebar while browsing away from viewer and initializing the same on returning back to the viewer. The state is saved in persistent store preferences and used upon viewer initialization.

Fixes https://github.com/mozilla/pdf.js/issues/6935

timvandermeij commented on web/viewer.js:
```diff
@@ -1061,9 +1063,13 @@ var PDFViewerApplication = {
+  setInitialView: function pdfViewSetInitialView(storedHash, scale, sidebar) {
     this.isInitialViewSet = true;
 
+    if(sidebar) {
```
Nit: add a space after `if`: `if (sidebar) {`

ankitaggarwal011 replied:
Done.

timvandermeij commented on web/viewer.js:
```diff
@@ -1061,9 +1063,13 @@ var PDFViewerApplication = {
     this.isInitialViewSet = true;
 
+    if(sidebar) {
+        document.getElementById('sidebarToggle').click();
```
Nit: indent with two spaces instead of four

ankitaggarwal011 replied:
Done.

timvandermeij commented:
@Snuffleupagus With the rotation persistance PR also in the queue, I'm a bit afraid that the `setInitialView` method becomes too cluttered with all the added logic and parameters. Do you have any idea if there is something we can do to prevent this? I'm not a big fan of adding a lot of parameters to methods: maybe pass in an object instead?

ankitaggarwal011 commented:
@timvandermeij, Thank you for the review. All the notes are taken care of and the changes have been pushed. 

> I'm a bit afraid that the ```setInitialView``` method becomes too cluttered with all the added logic and parameters.
This is exactly what I was thinking. I didn't make any changes in the flow of code since I was told to follow the rotation persistence PR, but passing these options through different function parameters might be a bad idea, as they the options tend to increase. 

> Do you have any idea if there is something we can do to prevent this?
I think there can be two solutions for this:
- The first one is the passing of a dictionary object in the ```setInitialView``` method with all the required values.
- The second one would be to simply skip passing any new parameter and directly use the ```store``` variable to get the desired values inside ```setInitialView``` method. However, this may hurt clarity.

What do you think? Let me know. Thanks.

timvandermeij commented:
#7038 just landed which refactored the sidebar code. With these patches in place, the code in this PR can be implemented as suggested in the diff at https://gist.github.com/Snuffleupagus/7aa031391399962846b0 (keep only the `sidebarView` stuff, not the rotation code).

Could you rebase this PR onto the current master branch and update the commit to be like the mentioned diff? Once it is like this, we should be able to land this soon. Thanks!

ankitaggarwal011 commented:
Sounds great. I'm on it :+1:

timvandermeij commented:
/botio-linux preview

ankitaggarwal011 commented:
Hi @timvandermeij, any update on this?

timvandermeij commented:
/botio-linux preview

timvandermeij commented on web/viewer.js:
```diff
@@ -917,7 +917,7 @@ var PDFViewerApplication = {
 
       store.initializedPromise.then(function resolved() {
-        var storedHash = null;
+        var storedHash = null, sidebarView;
```
It might be better to initialize this to `null` too.

ankitaggarwal011 replied:
Done.

timvandermeij commented on web/viewer.js:
```diff
@@ -1074,7 +1081,8 @@ var PDFViewerApplication = {
 
-    this.pdfSidebar.setInitialView(this.preferenceSidebarViewOnLoad);
+    this.pdfSidebar.setInitialView(this.preferenceSidebarViewOnLoad || 
+                                  (sidebarView | 0));
```
This line needs one extra space of indenting to align it with the `t` in `this.preferenceSidebarViewOnLoad`.

ankitaggarwal011 replied:
Done.

timvandermeij commented on web/viewer.js:
```diff
@@ -928,10 +928,14 @@ var PDFViewerApplication = {
           storedHash = 'page=1&zoom=' + self.preferenceDefaultZoomValue;
         }
-        self.setInitialView(storedHash, scale);
+        self.setInitialView(storedHash, { scale: scale,
```
Let's format this block as follows:

<pre>
        self.setInitialView(storedHash,
          { scale: scale, sidebarView: sidebarView });
</pre>

ankitaggarwal011 replied:
Done.

timvandermeij commented:
There is https://github.com/mozilla/pdf.js/blob/4d9a3d4fb9b667b25f62647de200a4816b2c903c/web/viewer.js#L961 that might need adjustment as well.

timvandermeij commented:
Sorry for the delayed review, I was quite busy the last few days. I did a review of this patch and added the comments above.

There appears to be another issue. Open the viewer from the preview, open `test/pdfs/basicapi.pdf` using the Open File button, open the sidebar, go to the outline view and refresh the page. The sidebar is indeed opened when the page reloads, but instead of the outline view the thumbnail view is shown.

ankitaggarwal011 commented:
It's alright. I understand. I will go through these comments/issues and let you know. I hope we will be able to close this soon. Thanks.

timvandermeij commented:
The code for this moved to `web/app.js`. Could you rebase the patch?

@Snuffleupagus I looked into the issue I mentioned above (about `test/pdfs/basicapi.pdf`) and I found that https://github.com/mozilla/pdf.js/pull/7010/files#diff-dd42e9b2d2a652b8e312efa567698aa6R1704 is 1 even though the outline view has value 2, hence the sidebar is not restored properly. The strange thing is that when the sidebar view is toggled, the value of `visibleView` is correctly updated, but at this point when we request it we still get the old value. Do you have any idea what could cause that, since you worked with the sidebar code before? It would be nice to land this patch, but this issue seems to be the only thing blocking that.

ankitaggarwal011 committed:
web/viewer.js: Persist the state of sidebar

Persist the state of content sidebar while browsing away from viewer and
initializing the same on returning back to the viewer. The state is saved
in persistent store preferences and used upon viewer initialization.

Fixes #6935

ankitaggarwal011 commented:
I have rebased the patch with the master. Phew! I hope we would be able to close this soon.

timvandermeij commented:
Correct. Fortunately there is #7190 now that should resolve that, so once that has landed this will probably be fine too. For now this is blocked by #7190.

ankitaggarwal011 commented:
Great. I'm glad that we got this figured out. I hope we can close this as soon as #7190 gets merged. Please let me know. Thanks.

timvandermeij commented:
Merged in #7190. Thank you for your contribution!

ankitaggarwal011 commented:
Cool. Great work! :+1:
